src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs

```diff
@@ -155,7 +155,7 @@ internal override void GenerateMethodBody(TypeCompilationState compilationState,
                 }
 
                 fields.Free();
-
+                retExpr = F.LogicalOr(F.ObjectEqual(F.This(), other), retExpr);
```

Comment:
>F.ObjectEqual(F.This(), other) [](start = 38, length = 30)

Is this check duplicated in the equality operator now? Should it be removed from there?

Reply:
@AlekseyTs The issue is that users can re-define `Equals`.

Reply:
> The issue is that users can re-define Equals.

And? The redefined Equals should still do the right thing.

---
In reply to: [574776221](https://github.com/dotnet/roslyn/pull/51149#discussion_r574776221) [](ancestors = 574776221)

Reply:
@AlekseyTs Just tried to remove the check from `==` implementation, and now null comparison fails unfortunately :confused:

The *current* implementation looks like:

```csharp
        if ((object)r1 != r2) // if this is removed, (MyRecord)null == (MyRecord)null be false.
        {
            if ((object)r1 != null)
            {
                return r1.Equals(r2);
            }
            return false;
        }
        return true;
```

This was caught by the following test:

```csharp
        [Theory]
        [CombinatorialData]
        public void EqualityOperators_09(bool useImageReference)
        {
            var source1 =
@"
public record A(int X) 
{
}
";
            var comp1 = CreateCompilation(source1);

            var source2 =
@"
class Program
{
    static void Main()
    {
        Test(null, null);
        Test(null, new A(0));
        Test(new A(1), new A(1));
        Test(new A(2), new A(3));
    }

    static void Test(A a1, A a2)
    {
        System.Console.WriteLine(""{0} {1} {2} {3}"", a1 == a2, a2 == a1, a1 != a2, a2 != a1);
    }
}
";
            CompileAndVerify(source2, references: new[] { useImageReference ? comp1.EmitToImageReference() : comp1.ToMetadataReference() }, expectedOutput:
@"
True True False False
False False True True
True True False False
False False True True
").VerifyDiagnostics();
        }
```

Reply:
> Just tried to remove the check from == implementation, and now null comparison fails unfortunately

I guess we could change the code gen to the following:
```
         if ((object)r1 != null)
            {
                return r1.Equals(r2);
            }

        return (object)r2 == null;
```

However, I am not sure if this is going to give us any noticeable improvement. 

---
In reply to: [575128469](https://github.com/dotnet/roslyn/pull/51149#discussion_r575128469) [](ancestors = 575128469)

Reply:
@dotnet/roslyn-compiler Thoughts? 

---
In reply to: [575311021](https://github.com/dotnet/roslyn/pull/51149#discussion_r575311021) [](ancestors = 575311021,575128469)

Reply:
> Is this check duplicated in the equality operator now? Should it be removed from there?

In the email thread, I believe we agreed to duplicate the equality check.